### ISSUE TITLE
Default transformRequest implementation for ArrayBufferView sends more data than intended

### ISSUE THREAD

**Issue by That3Percent** [axios/axios#2564]
The default [transformRequest ](https://github.com/axios/axios/blob/1a32ca0601850dd79d05bbec497c7fd3f6fcc346/lib/defaults.js#L43) implementation currently returns data.buffer for any ArrayBuffer view. If a view has an offset into that buffer, that offset is ignored.

For my use case, I'm serializing data using webassembly, and calling axios with a Uint8Array view of the data. Because this view is over a region of the webassembly heap, axios then tries to send all of the webassembly heap, rather than just the portion viewed by the Uint8Array.

I'm inclined to say this is a bug in axios, but I'm raising this as a question because it's difficult to say how to fix the bug without running into a backwards compatibility issue. If transformRequest could return a Uint8Array as a final step, and this function then created a matching Uint8Array view over the same ArrayBuffer as any given ArrayBufferView, that would solve the issue but would break anything which relied on the current behavior. I imagine it would be highly unusual to rely on the behavior as is though, and that most users of ArrayBuffer would expect it to respect the offset.

I hope to transform this issue into either a bug report or feature request with your guidance.

**Comment by yasuf** [comment]
couldn't you use `Uint8Array` by using a custom `transformRequest` ? you can pass any callback to the config in `transformRequest` and it will transform it in any way you need it to be transformed, that's just the default function that axios has, let me know if this helps

**Comment by That3Percent** [comment]
@yasuf According to the docs [here](https://github.com/axios/axios)
> // `transformRequest` allows changes to the request data before it is sent to the server
> // This is only applicable for request methods 'PUT', 'POST', 'PATCH' and 'DELETE'
> // The last function in the array must return a string or an instance of Buffer, ArrayBuffer,
> // FormData or Stream

Because the terminal value must be an ArrayBuffer and not a Uint8Array, one would need to copy the data to a new ArrayBuffer. It's possible, but surprising, unwieldy, and comes at a loss of performance.

**Comment by andrewparmet** [comment]
I just ran into this issue as well - it was very surprising and I did not expect this behavior. I would classify this as a bug myself, since the length of the sent data is not the length of the buffer I tried to send. At the very least this falls under principle of least surprise territory.

As for backwards compatibility, this project is still in the zero major version, so this kind of change can probably be done?
